### ISSUE TITLE
Add backlink panel

### ISSUE THREAD

**Issue by pomdtr** [kortina/vscode-markdown-notes#13]
**Describe the solution you'd like**
Beeing able to see which page link to the current page using a sidebar panel, as in roam research

**Describe alternatives you've considered**
Roam research, Obsidian, but they don't implement kakoune like keybindings + I love vscode

**Comment by kortina** [comment]
Yea, this is a cool idea. I was actually messing with a ripgrep + fzf version of this in bash: 

```sh
rg "\[\[.*md\]\]" --vimgrep --color always | fzf --ansi | rg -o "^.*:\d+:\d+" | xargs -I {} code -g {}
```
and was thinking about making a command to run this for the current file in the vscode terminal, but if there is a more native way to do this with the vscode sidebar panel, I think it would be an awesome addition.

Does anywhere here know what the options are for integration with the panel and which one would be the best for this?

**Comment by pomdtr** [comment]
I think we should for https://github.com/microsoft/vscode/issues/59921 to be finalised, in order to use the native vscode search (based on ripgrep) to find backlinks.

**Comment by pomdtr** [comment]
The sidebar can be customized with this api : https://code.visualstudio.com/api/extension-guides/tree-view

I also think that codelens could be usefull in order to see the number of reference for a file.

**Comment by pomdtr** [comment]
We can maybe use the built-in command : https://stackoverflow.com/questions/44740379/find-in-files-with-vs-code-extension

I propose to use a codelens at the top of the file, with a `See References/Unlinked` label.

**Comment by lukesmurray** [comment]
im curious if references solve this better than a custom tree view. I looked into the tree view api for embedding a view in the explorer panel and it seems like you need to manually refresh and you don't have access to the currently focused editor. Ideally I would want to create something like the outline view with a list of back references but that isn't exposed to the extension api as far as I can tell.

References may be better. Each reference is a back link. You can overload `gD` in vscode vim to be `show references`. which is a nice counter part to `gd` for go to definition.

![image](https://user-images.githubusercontent.com/34020210/80931807-39073b80-8d8a-11ea-8db1-04268e83a606.png)

**Comment by lukesmurray** [comment]
Also references can be transcluded using peek references. Each reference (except one) is a back link. One reference is the note itself which should be evident from the title.

![image](https://user-images.githubusercontent.com/34020210/80993368-f5074b80-8e08-11ea-9f17-7c4e319333a4.png)

**Comment by b3u** [comment]
I made an extension for this at https://github.com/b3u/vscode-backlinks-panel.

**Comment by lukesmurray** [comment]
@kortina I wasn't being creative enough when I read the API. @b3u shows a way to do it by listening for the change in visible text editors and then refreshing the tree view. 
```ts
    vscode.window.onDidChangeVisibleTextEditors(() =>
        provider.refresh()
    );
```

BTW @b3u you may want to switch to the `onDidChangeActiveTextEditor` event so the tree view represents the currently focused editor.

**Comment by b3u** [comment]
> BTW @b3u you may want to switch to the `onDidChangeActiveTextEditor` event so the tree view represents the currently focused editor.
@lukesmurray Thanks. I had just noticed this bug because I had two editor groups open.

@kortina Do you want me to integrate my code with this extension?

**Comment by kortina** [comment]
@lukesmurray ah, nice.

@b3u ya, for sure! That would be dope. If you pull request, do so off of https://github.com/kortina/vscode-markdown-notes/pull/19

Question for you guys: is the canonical name for a `[[wiki-link.md]]` that points to the file `wiki-link.md` a **Backlink**?  I have been calling them **References** / **Note References** (following more of the VS Code API naming convention), but if in the context of wikis/notebooks everyone calls them "backlinks" I guess that is what we should do in this project. 

This is nearly complete / should merge in soon / and I don't want you to get conflicts if you do a PR off of master.

**Comment by b3u** [comment]
Now that I think about it, having wikilinks *and* a backlinks panel might be more effective as an extension pack rather than one large extension. That way, someone could *just* add the wiki-style link syntax to markdown. What are your thoughts?

**Comment by kortina** [comment]
@b3u really your call -- I am planning to add a mix of related notes based on references and tags in the panel in the Tree View anyway, and think your stuff would be a great fit / help, and I'm happy to add you as one of the authors listed on the VS Code Marketplace page, but totally get it if you want to maintain a separate project, too.

My main thinking is that in working on this so far, I have seen a lot hinge upon the syntax for the wiki-links and tags, the regexes for defining them / allowable characters, and some of the configuration options, stuff that affects not only the syntax highlighting, but also the Reference and Definition finding, and I think the configuration for this stuff needs to be all in one place otherwise stuff will break / behave unexpectedly.

Just lmk after you think about it and we can talk about best way to integrate stuff if you decide you want to combine effort.

From my perspective, it would def be pretty cool having others involved in this project.

**Comment by lukesmurray** [comment]
>My main thinking is that in working on this so far, I have seen a lot hinge upon the syntax for the wiki-links and tags, the regexes for defining them / allowable characters, and some of the configuration options, stuff that affects not only the syntax highlighting, but also the Reference and Definition finding.

@kortina have you thought of using [remark](https://github.com/remarkjs/remark/tree/master/packages/remark-parse). It's a really nice system for parsing markdown into abstract syntax trees. It powers the gatsby blogging framework and has [plugin support for wiki links](https://github.com/landakram/remark-wiki-link#readme).

**Comment by kortina** [comment]
@lukesmurray you know, when I started working on this, I thought the regex matching would be simple enough, but I def should have thought about an AST parser when I started doing stuff like ReferenceProvider implementation. I'll check out that remark plugin for sure -- was not aware of it.

**Comment by kortina** [comment]
Closed by #29
